packages/connect-node/src/private/node-universal-client.ts

```diff
@@ -0,0 +1,290 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as http2 from "http2";
+import * as http from "http";
+import * as https from "https";
+import {
+  AsyncIterableSink,
+  Code,
+  ConnectError,
+  pipeTo,
+} from "@bufbuild/connect-core";
+import type {
+  UniversalClientFn,
+  UniversalClientRequest,
+  UniversalClientResponse,
+} from "./universal.js";
+import {
+  nodeHeaderToWebHeader,
+  webHeaderToNodeHeaders,
+} from "./node-universal-header.js";
+
+/**
+ * Create an HTTP client using the Node.js `http` or `https` package.
+ *
+ * The HTTP client is a simple function conforming to the type UniversalClientFn.
+ * It takes an UniversalClientRequest as an argument, and returns a promise for
+ * an UniversalClientResponse.
+ */
+export function createNodeHttp1Client(
+  httpOptions:
+    | Omit<http.RequestOptions, "signal">
+    | Omit<https.RequestOptions, "signal">
+    | undefined
+): UniversalClientFn {
+  return async function request(
+    req: UniversalClientRequest
+  ): Promise<UniversalClientResponse> {
+    const nodeRequest: http.ClientRequest = await new Promise(
+      (resolve, error) => {
+        let request: http.ClientRequest;
+        const requestOptions = {
+          ...httpOptions,
+          headers: webHeaderToNodeHeaders(req.header),
+          method: req.method,
+          signal: req.signal,
+        };
+        if (new URL(req.url).protocol.startsWith("https")) {
+          request = https.request(req.url, requestOptions);
+        } else {
+          request = http.request(req.url, requestOptions);
+        }
+        request.once("socket", (socket) => {
+          socket.once("connect", connect);
+          socket.once("error", error);
+
+          function connect() {
+            socket.off("error", rejectError);
+            resolve(request);
+          }
+
+          function rejectError(err: Error) {
+            socket.off("connect", connect);
+            error(err);
+          }
+        });
```

Comment:
@gilwong00, this ports https://github.com/bufbuild/connect-web/pull/390 - do you spot any obvious issues? (I know we are not converting errors yet)

Reply:
nope, this looks good!